validation: make `CCoinsView` a pure virtual interface

l0rinc opened the pull request:
### Problem
`CCoinsView` is the coins view interface, but historically it also provided built-in no-op behavior:

* It provided default no-op implementations (returning `std::nullopt`, `uint256()`, `false`, or `nullptr`) instead of being pure virtual.
* Callers could instantiate a bare `CCoinsView` and get silent no-op behavior.
* Mixing the interface definition with a built-in dummy implementation blurred the abstraction boundary.

### Context
This is part of the ongoing coins caching cleanup in #34280.

### Fix
This PR separates the interface from no-op behavior and makes `CCoinsView` pure virtual in incremental steps:
* Add `CoinsViewEmpty` as an explicit no-op coins view for tests, benchmarks, and temporary backends.
* Replace direct bare-`CCoinsView` test and dummy instantiations with `CoinsViewEmpty`.
* Make all `CCoinsView` methods pure virtual (`PeekCoin`, `GetCoin`, `HaveCoin`, `GetBestBlock`, `GetHeadBlocks`, `BatchWrite`, `Cursor`, `EstimateSize`).
* Remove the legacy default implementations from `coins.cpp`.
* Update fuzz and dummy backends to use `CoinsViewEmpty` explicitly.

ajtowns reviewed (commented):
Concept ACK, this seemed better separated as an interface and a dummy implementation to me too.

I think squashing all the commits after the first into a single commit would be better though.

andrewtoth commented on src/coins.cpp:
```diff
@@ -13,31 +13,11 @@ TRACEPOINT_SEMAPHORE(utxocache, add);
     CCoinsViewBacked(baseIn), m_deterministic(deterministic),
     cacheCoins(0, SaltedOutpointHasher(/*deterministic=*/deterministic), CCoinsMap::key_equal{}, &m_cache_coins_memory_resource)
 {
+    (void)base->GetBestBlock(); // Sanity check validity of base
```
I'm not sure about this sanity check. We're checking that a call to this method will work? Is this better than an assertion that baseIn is non-null? Why not refactor `baseIn` to be a reference?

l0rinc replied:
it's not null, it will give illegal access since it points to a random address. Try reverting and running any of the mentioned tests from the commit message.

andrewtoth replied:
Ok, but if it's a pointer it is legal for it to be a nullptr. So we should check first if it is not nullptr.
Also, instead of accessing a method, we could do a dynamic cast to `CCoinsView*` and assert the result.

andrewtoth replied:
Hmm dynamic casting uninitialized memory should be UB as well, but I guess it doesn't trigger a memory access violation.
But, we still need to check that `base` is not `nullptr` before accessing, otherwise passing `nullptr` would trigger undefined behavior.

l0rinc replied:
i don't think we support nullptr currently - it's why we have a default implementation with dummy values.

andrewtoth replied:
The API supports it. I use it in #31132. If we want to disallow it then we should change the parameter to a reference.

sipa replied:
As I understand it, this is sanity testing through triggering UB in the case the bug is present. I don't think that is ideal, as the compiler is allowed to assume it won't happen and optimize it away. This isn't too far fetched - if it could be inferred that (all implementations of) `CCoinsView::GetBestBlock` are side-effect free, then the statement `(void)base->GetBestBlock();` could just be optimized away.

Testing for bugs due to initialization order seems more like a job for UBSan. Unfortunately, it seems none of the current fuzz tests (with UBSan enabled) trigger with the following patch (on master):

```diff
diff --git a/src/coins.cpp b/src/coins.cpp
index 14381ecb0aa..c7206c22554 100644
--- a/src/coins.cpp
+++ b/src/coins.cpp
@@ -38,6 +38,7 @@ CCoinsViewCache::CCoinsViewCache(CCoinsView* baseIn, bool deterministic) :
     CCoinsViewBacked(baseIn), m_deterministic(deterministic),
     cacheCoins(0, SaltedOutpointHasher(/*deterministic=*/deterministic), CCoinsMap::key_equal{}, &m_cache_coins_memory_resource)
 {
+    assert(!base || dynamic_cast<CCoinsView*>(base));
     m_sentinel.second.SelfRef(m_sentinel);
 }
```

I'm not objecting to this change, but I think it would be better if we could test for bugs of this class in a more systematic way.

andrewtoth replied:
> I needed something to demonstrate the UB - and that it's fixed now

Would it be correct to say there is no UB then? UB is only triggered by dereferencing the pointer to uninitialized memory, which master does not do?

l0rinc replied:
The reason is explain in the commit message:
> In a follow-up `m_dummy` will be completely removed, this commit was meant to demonstrate why that was necessary.

I.e. I don't want the fix to be accidental - fix first, replace after.

andrewtoth replied:
But it's not necessary. There is no UB unless you add this commit.

l0rinc commented on src/coins.h:
```diff
@@ -378,6 +403,7 @@ class CCoinsViewCache : public CCoinsViewBacked
 
 public:
     CCoinsViewCache(CCoinsView *baseIn, bool deterministic = false);
+    CCoinsViewCache() : CCoinsViewCache{&CCoinsViewEmpty::Get()} {}
```
How so? We have to init `m_deterministic` as well...

sipa reviewed (commented):
Concept ACK

andrewtoth commented on src/coins.h:
```diff
@@ -303,58 +303,83 @@ struct CoinsViewCacheCursor
     //! Do a bulk modification (multiple Coin changes + BestBlock change).
     //! The passed cursor is used to iterate through the coins.
-    virtual bool BatchWrite(CoinsViewCacheCursor& cursor, const uint256& hashBlock);
+    virtual bool BatchWrite(CoinsViewCacheCursor& cursor, const uint256& hashBlock) = 0;
```
```suggestion
    virtual bool BatchWrite(CoinsViewCacheCursor& cursor, const uint256& block_hash) = 0;
```

andrewtoth commented on src/coins.h:
```diff
@@ -303,58 +303,83 @@ struct CoinsViewCacheCursor
+    bool HaveCoin(const COutPoint&) const override { return false; }
+    uint256 GetBestBlock() const override { return {}; }
+    std::vector<uint256> GetHeadBlocks() const override { return {}; }
+    bool BatchWrite(CoinsViewCacheCursor&, const uint256&) override { return false; }
```
Should we iterate through the cursor here to clear it, and return `true`? If we return `false` we leave the parent cache in an inconsistent state.

andrewtoth commented on src/coins.h:
```diff
@@ -303,58 +303,83 @@ struct CoinsViewCacheCursor
-    std::unique_ptr<CCoinsViewCursor> Cursor() const override;
-    size_t EstimateSize() const override;
-};
+    explicit CCoinsViewBacked(CCoinsView* viewIn) : base{Assert(viewIn)} {}
```
```suggestion
    explicit CCoinsViewBacked(CCoinsView* view_in) : base{Assert(view_in)} {}
```

l0rinc replied:
Done

l0rinc commented on src/test/fuzz/coins_view.cpp:
```diff
@@ -93,10 +93,11 @@ void TestCoinsView(FuzzedDataProvider& fuzzed_data_provider, CCoinsView& backend
             [&] {
-                if (fuzzed_data_provider.ConsumeBool()) {
-                    backend_coins_view = CCoinsView{};
+                if (is_db || fuzzed_data_provider.ConsumeBool()) {
```
Can you explain why you think so? If we have a database backing the cache than we should switch it to another one? in https://github.com/bitcoin/bitcoin/blob/7817197cf2eb07aae32a6f67408c4493a57ea465/src/validation.cpp#L880 we have a normal db switched to an empty one. Here we're either setting the provided `backend_coins_view` or creating a new one - is that not what you think we should do?

l0rinc replied:
Hmmm, would this work instead?
```patch
diff --git a/src/test/fuzz/coins_view.cpp b/src/test/fuzz/coins_view.cpp
--- a/src/test/fuzz/coins_view.cpp	(revision 918befc2a2341221f0fbdccbd2c89f29751122b3)
+++ b/src/test/fuzz/coins_view.cpp	(revision c02ba80c7e5adaf16e6c2a22a2b2e4a6e38c007b)
@@ -42,11 +42,11 @@
     static const auto testing_setup = MakeNoLogFileContext<>();
 }
 
-void TestCoinsView(FuzzedDataProvider& fuzzed_data_provider, CCoinsView& backend_coins_view, bool is_db)
+void TestCoinsView(FuzzedDataProvider& fuzzed_data_provider, CCoinsView* backend_coins_view, bool is_db)
 {
     bool good_data{true};
 
-    CCoinsViewCache coins_view_cache{&backend_coins_view, /*deterministic=*/true};
+    CCoinsViewCache coins_view_cache{backend_coins_view, /*deterministic=*/true};
     if (is_db) coins_view_cache.SetBestBlock(uint256::ONE);
     COutPoint random_out_point;
     Coin random_coin;
@@ -94,9 +94,9 @@
             },
             [&] {
                 if (fuzzed_data_provider.ConsumeBool()) {
-                    backend_coins_view = CCoinsView{};
+                    backend_coins_view = &CCoinsViewEmpty::Get();
                 }
-                coins_view_cache.SetBackend(backend_coins_view);
+                coins_view_cache.SetBackend(*backend_coins_view);
             },
             [&] {
                 const std::optional<COutPoint> opt_out_point = ConsumeDeserializable<COutPoint>(fuzzed_data_provider);
@@ -176,11 +176,11 @@
             assert(!exists_using_access_coin && !exists_using_have_coin_in_cache && !exists_using_have_coin);
         }
         // If HaveCoin on the backend is true, it must also be on the cache if the coin wasn't spent.
-        const bool exists_using_have_coin_in_backend = backend_coins_view.HaveCoin(random_out_point);
+        const bool exists_using_have_coin_in_backend = backend_coins_view->HaveCoin(random_out_point);
         if (!coin_using_access_coin.IsSpent() && exists_using_have_coin_in_backend) {
             assert(exists_using_have_coin);
         }
-        if (auto coin{backend_coins_view.GetCoin(random_out_point)}) {
+        if (auto coin{backend_coins_view->GetCoin(random_out_point)}) {
             assert(exists_using_have_coin_in_backend);
             // Note we can't assert that `coin_using_get_coin == *coin` because the coin in
             // the cache may have been modified but not yet flushed.
@@ -206,11 +206,11 @@
     }
 
     {
-        std::unique_ptr<CCoinsViewCursor> coins_view_cursor = backend_coins_view.Cursor();
+        std::unique_ptr<CCoinsViewCursor> coins_view_cursor = backend_coins_view->Cursor();
         assert(is_db == !!coins_view_cursor);
-        (void)backend_coins_view.EstimateSize();
-        (void)backend_coins_view.GetBestBlock();
-        (void)backend_coins_view.GetHeadBlocks();
+        (void)backend_coins_view->EstimateSize();
+        (void)backend_coins_view->GetBestBlock();
+        (void)backend_coins_view->GetHeadBlocks();
     }
 
     if (fuzzed_data_provider.ConsumeBool()) {
@@ -297,8 +297,7 @@
 FUZZ_TARGET(coins_view, .init = initialize_coins_view)
 {
     FuzzedDataProvider fuzzed_data_provider{buffer.data(), buffer.size()};
-    CCoinsView backend_coins_view;
-    TestCoinsView(fuzzed_data_provider, backend_coins_view, /*is_db=*/false);
+    TestCoinsView(fuzzed_data_provider, &CCoinsViewEmpty::Get(), /*is_db=*/false);
 }
 
 FUZZ_TARGET(coins_view_db, .init = initialize_coins_view)
@@ -310,5 +309,5 @@
         .memory_only = true,
     };
     CCoinsViewDB coins_db{std::move(db_params), CoinsViewOptions{}};
-    TestCoinsView(fuzzed_data_provider, coins_db, /*is_db=*/true);
+    TestCoinsView(fuzzed_data_provider, &coins_db, /*is_db=*/true);
 }
```

andrewtoth replied:
Can we just keep the same code without checking `is_db` and just assigning the empty view now?
```C++
                if (fuzzed_data_provider.ConsumeBool()) {
                    backend_coins_view = CCoinsViewEmpty::Get();
                }
```

l0rinc commented on src/coins.h:
```diff
@@ -303,58 +303,83 @@ struct CoinsViewCacheCursor
+    bool HaveCoin(const COutPoint& outpoint) const override { return base->HaveCoin(outpoint); }
+    uint256 GetBestBlock() const override { return base->GetBestBlock(); }
+    std::vector<uint256> GetHeadBlocks() const override { return base->GetHeadBlocks(); }
+    bool BatchWrite(CoinsViewCacheCursor& cursor, const uint256& hashBlock) override { return base->BatchWrite(cursor, hashBlock); }
```
Separated the renames and simple moves to a completely new commit to simplify review

l0rinc commented:
A slightly more involved rebase after https://github.com/bitcoin/bitcoin/pull/33866.
Most of the conflicts were around moving `BatchWrite` around.
I have also separated the second commit into 3, to address reviewer feedback regarding cleaning up naming and formatting of affected methods. This way - after a boring refactor commit - the non-trivial follow-up commits are easier to review.
Can be reviewed via `git range-diff -w -U0 master e3d4435 d82eedd`.

w0xlt reviewed (commented):
ACK 3b800c9e939f3c8dcaef6537624033c5be01a64e

1) Good separation between CCoinsView (interface), CCoinsViewEmpty (no-op), and CCoinsViewBacked (delegator).
2) Test files are notably cleaner, removing the repetitive `CCoinsView coinsDummy; CCoinsViewCache coins(&coinsDummy);` pattern.

ryanofsky commented on src/coins.cpp:
```diff
@@ -43,7 +43,6 @@ CCoinsViewCache::CCoinsViewCache(CCoinsView* base_in, bool deterministic) :
     CCoinsViewBacked(base_in), m_deterministic(deterministic),
     cacheCoins(0, SaltedOutpointHasher(/*deterministic=*/deterministic), CCoinsMap::key_equal{}, &m_cache_coins_memory_resource)
 {
-    (void)base->GetBestBlock(); // Sanity check validity of base
```
In commit "refactor: add `CCoinsViewEmpty` as a noop implementation for base and testing" (9e08e4881ea02120573480f22bc2082dd8f57f6e)

Unclear why this line is being deleted this commit. Might be worth mentioning in commit message if it is was needed earlier but is redundant now, or something like that.

l0rinc replied:
Removed the commit, thanks.

ryanofsky reviewed (approved):
Code review 515dbf1e72ef4ba0233aedcc4b40f44221f899dc. This looks good but I want to spend a little more time checking before acking. Making `CoinsView` pure-virtual seems like a clearly good change. Some things here are maybe not ideal though:

- It doesn't seem good that `CCoinsViewCache` is now default constructible, but if it is default-constructed it behaves like a test class and discards data written to it. It would seem safer to requiring passing a no-op base view if you want that behavior, or to provide a separate `NoOpCoinsCache` subclass that discards data for tests.

- It seems unfortunate to lose the default `HaveCoin(outpoint)` implementation (`GetCoin(outpoint).has_value()`) and force subclasses to re-implement it. Not sure if there would be any downside to keeping this method.

In general removing the no-op behaviors from the `CCoinsView` base class seems like a clearly good thing, but it would be nice to reduce uses of CCoinsViewEmpty as well. Maybe a followup could do this (or some followups already do).

ryanofsky commented on src/validation.cpp:
```diff
@@ -874,10 +869,14 @@ bool MemPoolAccept::PreChecks(ATMPArgs& args, Workspace& ws)
+    //
+    // When m_view is connected to an empty view, it can no longer look up coins from the mempool
+    // or UTXO set (meaning no disk operations happen), but can still return coins it accessed
+    // previously. Useful for keeping track of which coins were pulled from disk.
```
In commit "refactor: add `CCoinsViewEmpty` as a noop implementation for base and testing" (ee76bc0056dfa094a467851bfb5f2fb23ac13b39)

This new comment is good but also makes be curious. Is it just theoretically useful for keeping track of which coins were pulled from disk, or is there code which actually does that? Might be good to clarify.

ryanofsky commented on src/test/coins_tests.cpp:
```diff
@@ -53,6 +53,7 @@ class CCoinsViewTest : public CCoinsView
     }
 
     uint256 GetBestBlock() const override { return hashBestBlock_; }
+    size_t EstimateSize() const override { return map_.size(); }
```
In commit "refactor: make `CCoinsView::EstimateSize` purely virtual" (dbebcbcdde2a44b9cd0b4b4032fedc62749ef54d)

It seems like it might be better to just throw not implemented or return 0 like before instead of returning a fake size. It looks like this method is supposed to return an estimation of disk space used by database, not a count of items in the map.

ryanofsky commented on src/coins.cpp:
```diff
@@ -14,14 +14,7 @@ TRACEPOINT_SEMAPHORE(utxocache, add);
 TRACEPOINT_SEMAPHORE(utxocache, spent);
 TRACEPOINT_SEMAPHORE(utxocache, uncache);
 
-std::optional<Coin> CCoinsView::GetCoin(const COutPoint& outpoint) const { return std::nullopt; }
```
re: https://github.com/bitcoin/bitcoin/pull/34124#discussion_r2774975617

> Let me know if this is what you meant.

Yes, that's exactly what I meant, thanks!

ryanofsky reviewed (approved):
Code review ACK 4fdcf62bb56c0549dfff1f812a6ce880c42f5989. Very nice changes, and I'm much more comfortable with them having gone through the PR a second time. Thanks for all the updates since last review! I left more suggestions below but none are important.

ajtowns commented on src/coins.cpp:
```diff
@@ -43,6 +43,7 @@ CCoinsViewCache::CCoinsViewCache(CCoinsView* baseIn, bool deterministic) :
     CCoinsViewBacked(baseIn), m_deterministic(deterministic),
     cacheCoins(0, SaltedOutpointHasher(/*deterministic=*/deterministic), CCoinsMap::key_equal{}, &m_cache_coins_memory_resource)
 {
+    (void)base->GetBestBlock(); // Sanity check validity of base
```
Yeah, afaics you're introducing the bug (derefencing the uninitialized base where previously it was just storing an address and not dereferencing until post-initialization) at the same time you're fixing it (getting the object initialized first). Not wrong, but seems over complicated.

l0rinc replied:
Thanks for the comments, now that the potential bug is clear, I agree that there's no point in triggering it in a separate commit, I have removed it, mentioning it only in the commit message.

ajtowns commented on src/coins.h:
```diff
@@ -330,6 +330,33 @@ class CCoinsView
+
+    std::optional<Coin> GetCoin(const COutPoint&) const override { return {}; }
+    bool HaveCoin(const COutPoint&) const override { return false; }
+    uint256 GetBestBlock() const override { return {}; }
```
This doesn't make sense to me -- `m_view` is explicitly a caching coins view, and changing the backend explicitly doesn't clear the cache (otherwise the utxos that were already loaded wouldn't go away). Calling `GetBestBlock()` after changing the backend and getting a cached result is completely reasonable behaviour here, and precisely consistent with how utxos are treated, it's not "fake", "lazy" or "super-smelly".

andrewtoth commented on src/wallet/wallet.cpp:
```diff
@@ -1324,7 +1324,7 @@ bool CWallet::AbandonTransaction(CWalletTx& tx)
 }
 
-void CWallet::MarkConflicted(const uint256& hashBlock, int conflicting_height, const Txid& hashTx)
+void CWallet::MarkConflicted(const uint256& hash_block, int conflicting_height, const Txid& hashTx)
```
nit: This doesn't seem like it's related? It's not a `CCoinsView` subclass. Don't really mind this being changed, but if touching up could either remove it or also update `hashTx` -> `tx_hash`.

andrewtoth commented on src/test/coins_tests.cpp:
```diff
@@ -54,7 +54,7 @@ class CCoinsViewTest : public CCoinsView
     uint256 GetBestBlock() const override { return hashBestBlock_; }
 
-    void BatchWrite(CoinsViewCacheCursor& cursor, const uint256& hashBlock) override
+    void BatchWrite(CoinsViewCacheCursor& cursor, const uint256& hash_block) override
```
nit: This parameter should actually be `block_hash` everywhere, since we're reformatting.

l0rinc replied:
If the reviewers don't mind - given it's already a cleanup PR - done!

andrewtoth commented on src/coins.cpp:
```diff
@@ -29,18 +29,18 @@ bool CCoinsView::HaveCoin(const COutPoint &outpoint) const
 }
 
-CCoinsViewBacked::CCoinsViewBacked(CCoinsView *viewIn) : base(viewIn) { }
+CCoinsViewBacked::CCoinsViewBacked(CCoinsView* view_in) : base(view_in) {}
```
nit: Should be `in_view` everywhere.

andrewtoth commented on src/test/fuzz/coins_view.cpp:
```diff
@@ -107,8 +107,8 @@ void TestCoinsView(FuzzedDataProvider& fuzzed_data_provider, CCoinsView* backend
             [&] {
-                if (fuzzed_data_provider.ConsumeBool()) {
-                    *backend_coins_view = CCoinsView{};
+                if (!is_db && fuzzed_data_provider.ConsumeBool()) {
```
I think we should remove the `!is_db` here. It is removing fuzz coverage.

Also, I don't think we should describe adding new fuzz paths as doing things "randomly". It is adding coverage. "The fuzzer was also updated to cover adding both a real db and empty backing view."

andrewtoth commented on src/validation.cpp:
```diff
@@ -878,7 +879,7 @@ bool MemPoolAccept::PreChecks(ATMPArgs& args, Workspace& ws)
     m_view.SetBackend(CCoinsViewEmpty::Get());
 
-    assert(m_active_chainstate.m_blockman.LookupBlockIndex(m_view.GetBestBlock()) == m_active_chainstate.m_chain.Tip());
+    assert(m_active_chainstate.m_blockman.LookupBlockIndex(best_block) == m_active_chainstate.m_chain.Tip());
```
I think we can drop this commit. I don't think it's related to the goal of this PR.

andrewtoth reviewed (approved):
ACK 4fdcf62bb56c0549dfff1f812a6ce880c42f5989

I think that there are some unnecessary changes that could be dropped and commits removed. The last commit just seems like code churn.

The commits that make individual methods pure virtual can be combined into one. I think splitting them up like that makes it harder to review.

ajtowns commented:
I don't think 8eb773f832b56ca0e0d9971d2bcc8f42f3abe465 (rearranging members that are getting deleted anyway) is very useful, and a1803c9a46225a8673b5ff3d23b389a18c38965e (doing a local cache to avoid using the CCoinsViewCache cache) seems like a step backwards. Weak preference for dropping d2e516e079bae495876b644a1ce4fef1fee0c09e (removing the default for HaveCoin and reimplementing it twice) as well. Still think squashing all the "make xxx purely virtual" commits into a single commit would also be an improvement. The fuzz changes confused me, but seem to be fixing potential UB here, so that's good, but should go further I think.

l0rinc committed:
fuzz: keep backend assertions aligned to active backend

`TestCoinsView` switches the `CCoinsViewCache` backend during fuzzing and then queries the backend for cross-checks.
Pass the backend as a `CCoinsView*` (to make it relocatable) and retarget it when toggling between the original backend and a local empty `CCoinsView` so assertions always refer to the active backend. This will be switched to a singleton in the next commit.

Note that the previous slice-assignment (`backend_coins_view = CCoinsView{}`) was a silent noop for the db target, it only copied base-class members (none) without changing the vtable, so the backend was never actually switched.
The pointer approach makes the switch real for both targets, which revealed that when restoring the original backend after the empty one, the cache must be reset first to avoid carrying FRESH flags that were valid relative to the empty backend but invalid relative to the original (which may already contain those coins).

Co-authored-by: Anthony Towns <aj@erisian.com.au>
Co-authored-by: Andrew Toth <andrewstoth@gmail.com>
Co-authored-by: Ryan Ofsky <ryan@ofsky.org>
Co-authored-by: marcofleon <marleo23@proton.me>

l0rinc committed:
coins: add explicit `CoinsViewEmpty` noop backend

 Introduce `CoinsViewEmpty` as an explicit no-op `CCoinsView` implementation, and define its singleton accessor out of line in `coins.cpp` to avoid `-Wunique-object-duplication` in shared-library builds.`
Use it at call sites that intentionally want a no-op backend instead of constructing anonymous placeholder views.

`CCoinsViewTest` and `CoinsViewBottom` now inherit defaults from `CoinsViewEmpty` (e.g. the unused `EstimateSize()`, which now returns 0).

Co-authored-by: Ryan Ofsky <ryan@ofsky.org>

l0rinc committed:
scripted-diff: normalize `CCoinsView` naming

Standardize coins view naming with a mechanical rename pass.
This keeps subsequent commits focused on the interface and behavioral changes.

Co-authored-by: Andrew Toth <andrewstoth@gmail.com>

-BEGIN VERIFY SCRIPT-
git grep -qE '\bin_base\b|\bin_view\b|\bin_block_hash\b|\bblock_hash\b' -- src/coins.cpp src/coins.h src/txdb.cpp src/txdb.h src/test/coins_tests.cpp && { echo "Error: target names already exist in scoped files"; exit 1; }

perl -pi -e '
  s/\bbaseIn\b/in_base/g;
  s/\bhashBlockIn\b/in_block_hash/g;
  s/\bhashBlock\b/block_hash/g;
  s/\bviewIn\b/in_view/g;
  ' src/coins.cpp src/coins.h src/txdb.cpp src/txdb.h src/test/coins_tests.cpp
-END VERIFY SCRIPT-

Eunovo reviewed (commented):
ACK https://github.com/bitcoin/bitcoin/pull/34124/commits/877e713dadea535adbc0e3e79bfe106ad1c2bbff

ryanofsky reviewed (approved):
Code review ACK 877e713dadea535adbc0e3e79bfe106ad1c2bbff. Seems to need rebase due to #33512 but looks great! Lots of nice simplifications since last review.

w0xlt reviewed (commented):
re-ACK 356eb7cd95456d79fbec806ee2105e70b4399a96

Ci error is unrelated.

l0rinc committed:
refactor: normalize CCoinsView whitespace and signatures

Let's get these out of the way to simplify riskier followup commits

l0rinc committed:
coins: make `CCoinsView` methods pure virtual

`CCoinsView` provided default no-op implementations, which allowed constructing a bare view and silently getting dummy behavior.
Make all interface methods pure virtual and remove the legacy default definitions from `coins.cpp` so callers must choose an explicit implementation.
Move the virtual destructor to the beginning to avoid mixing it between the methods.
No-op backing behavior remains available via `CoinsViewEmpty`.

l0rinc committed:
refactor: inline `CCoinsViewBacked` implementation

`CCoinsViewBacked` is a simple delegating wrapper around another `CCoinsView`.

Inline its one-line overrides in `coins.h` so the view hierarchy can be read without jumping between `coins.h` and `coins.cpp`.

l0rinc committed:
refactor: rename `hashBlock` to `m_block_hash` to avoid shadowing

w0xlt reviewed (commented):
reACK 7d0971835abe303e84a7f4b00663665a373181b0

Eunovo reviewed (commented):
ReACK https://github.com/bitcoin/bitcoin/pull/34124/commits/7d0971835abe303e84a7f4b00663665a373181b0

ryanofsky commented on src/coins.cpp:
```diff
@@ -27,12 +27,12 @@ std::unique_ptr<CCoinsViewCursor> CCoinsView::Cursor() const { return nullptr; }
 CCoinsViewBacked::CCoinsViewBacked(CCoinsView* in_view) : base(in_view) { }
-std::optional<Coin> CCoinsViewBacked::GetCoin(const COutPoint& outpoint) const { return base->GetCoin(outpoint); }
 std::optional<Coin> CCoinsViewBacked::PeekCoin(const COutPoint& outpoint) const { return base->PeekCoin(outpoint); }
+std::optional<Coin> CCoinsViewBacked::GetCoin(const COutPoint& outpoint) const { return base->GetCoin(outpoint); }
```
In commit "coins: route default coin lookups through `PeekCoin`" (fcdd08035c27db1414cdc7b3fadbe4481e5ec5cd)

There are a lot of move-only changes in this commit (this one and 7 others) where code seems to be changing position for no clear reason. These changes are difficult to review even with --color-moved, because in addition to the moves, there is also the line `std::optional<Coin> PeekCoin(const COutPoint& outpoint) const override;` that is copied a few times and also shown in grey, so it's difficult to distinguish the copies from the moves. Would suggest just dropping the moves from this PR to keep it simple, or if keeping the moves, doing it in a separate commit and saying what intention behind them is in the commit message.

l0rinc replied:
Yeah, the method order should be a separate refactor, thanks, reverted the order.

ryanofsky reviewed (approved):
Code review 7d0971835abe303e84a7f4b00663665a373181b0. Looks very good, but I have some concerns about "route default coin lookups" commit and suggested some simple changes that would resolve them (see below)

ryanofsky commented on src/coins.cpp:
```diff
@@ -49,6 +30,14 @@ std::optional<Coin> CCoinsViewCache::FetchCoinFromBase(const COutPoint& outpoint
     return base->GetCoin(outpoint);
 }
 
+std::optional<Coin> CCoinsViewCache::PeekCoin(const COutPoint& outpoint) const
```
In commit "coins: route default coin lookups through `PeekCoin`" (76db070f78dd93710f722523f5dd14fbea80efe5)

There seems to be an unnecessary move-only change here. (Also noticed an unnecessary destructor move in 76db070f78dd93710f722523f5dd14fbea80efe5). As mentioned before I'd probably drop these moves because they are distracting, but if keeping them it'd be good to mention some reasoning in the commit message. For example this `PeekCoin` definition is being placed between `FetchCoinFromBase` and `FetchCoin`, which seems random, so could help to know why it's moving and why it's moving there.

ryanofsky reviewed (approved):
Code review ACK fa48fed58c607578f68719282a903e377d5e63a1. Thanks for the updates! PR now seems simple and safer with the latest "route default coin lookups through `PeekCoin`" update. The new `CCoinsViewEmpty` class is also nice for eliminating dummy variable everywhere and making `CCoinsView` safer to inherit from.

ryanofsky commented on src/test/fuzz/coins_view.cpp:
```diff
@@ -255,13 +255,12 @@ void TestCoinsView(FuzzedDataProvider& fuzzed_data_provider, CCoinsViewCache& co
-            std::unique_ptr<CCoinsViewCursor> coins_view_cursor = backend_coins_view.Cursor();
-            assert(!!coins_view_cursor);
+        if (!dynamic_cast<CCoinsViewCache*>(backend_coins_view)) {
+            assert(!!backend_coins_view->Cursor() == (dynamic_cast<CCoinsViewDB*>(backend_coins_view) != nullptr));
```
re: https://github.com/bitcoin/bitcoin/pull/34124#discussion_r2863931858

> Does the above work for you though? I get an instant failure:
> 
> ```shell
> Assertion failed: (!!backend_coins_view->Cursor() == (is_db && backend_coins_view == backend_coins_view)), function TestCoinsView, file coins_view.cpp, line 236.
> ==70815== ERROR: libFuzzer: deadly signal
> ```

It looks like there is a typo here and `backend_coins_view == backend_coins_view` should be replaced with `backend_coins_view == original_backend`. I haven't tried running the fuzzer at all though and current version seems ok, even if the check is less strict than it could be.

l0rinc replied:
Changed, thanks.

ryanofsky reviewed (approved):
Code review ACK b91e279fe904c7a2808245123563b19f4106698e just cleaning up move-only changes since last review

andrewtoth commented on src/txmempool.cpp:
Here we are calling `base->GetCoin()` from `PeekCoin()`. Should we make this `base->PeekCoin()` instead here? What are the implications of caching here? I would think we want to cache when looking up coins in `CCoinsViewMemPool` for performance reasons. Perhaps we don't change this method to `PeekCoin`, and instead write a new `CCoinsViewMemPool::PeekCoin`?

ryanofsky replied:
re: https://github.com/bitcoin/bitcoin/pull/34124#discussion_r2887123554

> Here we are calling `base->GetCoin()` from `PeekCoin()`. Should we make this `base->PeekCoin()` instead here? What are the implications of caching here? I would think we want to cache when looking up coins in `CCoinsViewMemPool` for performance reasons. Perhaps we don't change this method to `PeekCoin`, and instead write a new `CCoinsViewMemPool::PeekCoin`?

Nice catch. Yeah I definitely think it makes sense for `CCoinsViewMemPool::PeekCoin` to call `base->PeekCoin` and the GetCoin implementation to call GetCoin. Or for PeekCoin not to be implemented and throw an exception.

This [again](https://github.com/bitcoin/bitcoin/pull/34165#discussion_r2789324653) makes me think that defining PeekCoin GetCoin and HaveCoin in terms of each other and having subclasses override them is inherently unsafe because there is no guarantee subclasses will implement all of the methods if they implement some. In longer run, may want to make code look more like suggestion in the linked comment where subclasses only override pure virtual hooks without default implementations, and the top-level class can provide convenient non-virtual accessors on top of those hooks.

ryanofsky commented on src/txmempool.cpp:
```diff
@@ -759,9 +759,21 @@ std::optional<Coin> CCoinsViewMemPool::GetCoin(const COutPoint& outpoint) const
+std::optional<Coin> CCoinsViewMemPool::GetCoin(const COutPoint& outpoint) const
+{
+    if (auto coin{GetCoinFromMemPool(outpoint)}) return coin;
     return base->GetCoin(outpoint);
```
In commit "coins: route default coin lookups through `PeekCoin`" (8bd491237b3eba3f332a7a2d1e46c800b406f2d0)

It seems like there is now a slight change in behavior here because in the first case where `GetCoinFromMemPool` returns nullopt on line 760, when the output contains a valid transaction id but an invalid output index, a base lookup happen when there was no base lookup previously.

I guess this is ok because the transaction shouldn't exist in the base view, so return value will ultimately be nullopt, but if keeping this change it would be good to mention it in the commit message.

I'd probably avoid changing behavior, though, by having the `GetCoinFromMemPool` return different values in the two cases instead of `std::nullopt` in both cases. Could have it return something like `tuple<bool, optional<Coin>>` or `variant<monostate, optional<Coin>>` or just add a bool output parameter.

ryanofsky reviewed (approved):
Code review ACK 39cdec5b05fb1e872e2cca67dd4ee4c3ccf08740. Since last review just renamed `CoinsViewEmpty` and droped the change which tried to make `PeekCoin`, which is not supposed to modify caches, avoid calling `GetCoin`, which does modify them. Agree it's best to drop that change and not try to implement it incorrectly for the nth time.

andrewtoth reviewed (approved):
ACK f69fc9ed9109cd844f8b25ac63a6bc9024f47fa3

PeekCoin delegating to GetCoin in CCoinsViewDB and CCoinsViewMemPool is not ideal, but doesn't change current behavior. So, improving this can be left to a follow-up.

ryanofsky reviewed (approved):
Code review ACK f69fc9ed9109cd844f8b25ac63a6bc9024f47fa3. Just comment changes since past review

re: https://github.com/bitcoin/bitcoin/pull/34124#pullrequestreview-3912322099

> PeekCoin delegating to GetCoin in CCoinsViewDB and CCoinsViewMemPool is not ideal, but doesn't change current behavior. So, improving this can be left to a follow-up.

Agree with this, but just to be clear in the case of `CCoinsViewMemPool::PeekCoin` there is a local change in behavior because previously calling `PeekCoin` would return coins from the base view, and now it will call `GetCoin` instead.. There is just no overall change in behavior just because the mempool peekcoin method is not currently called.

w0xlt reviewed (commented):
reACK https://github.com/bitcoin/bitcoin/commit/f69fc9ed9109cd844f8b25ac63a6bc9024f47fa3

ajtowns reviewed (commented):
NACK f69fc9ed9109cd844f8b25ac63a6bc9024f47fa3

I think the inheritable default implementations of HaveCoin and PeekCoin would be better in the base class than duplicated across CoinsViewEmpty and CCoinsViewDB, but particularly with PeekCoin rework planned, nbd.

EDIT to add: Changing the `CCoinsViewMemPool::PeekCoin` behaviour in a refactor is not okay.

There was a lot of things that looked to me like they went a few steps beyond refactoring in various earlier iterations of this PR; perhaps worth a re-look at CONTRIBUTING.md comments on refactoring PRs ("`refactor` for structural changes that do not change behavior" and "In all cases, refactoring PRs must not change the behaviour of code within the pull request (bugs must be preserved as is)." eg)

sedited reviewed (commented):
Concept ACK

Might it be better to squash "coins: make `CCoinsView` methods pure virtual" into "coins: move default coin lookups to `CoinsViewEmpty`"? I think that would make the mechanics of "we need to add this override, because its implementation is no longer available" a bit clearer. I'm also not sure if we can really call this a refactor. Maybe prefix the PR title with `validation:` instead?

l0rinc reviewed (commented):
Thanks for the reviews.
I reverted the contentious `CCoinsViewMemPool::PeekCoin` fix, changed the title to not mention `refactor`, squashed two commits doing similar things, and made the default empty view constructor protected.

ryanofsky reviewed (approved):
Code review ACK 1d1dc4ed2cfcc573e73ddef2d15b6e64df4f5ff3. Since [last review](https://github.com/bitcoin/bitcoin/pull/34124#pullrequestreview-3916864785), PR was rebased, `CoinsViewEmpty` constructor was made protected, pure virtual methods were moved in a single commit instead of two commits, and the new `CCoinsViewMemPool::PeekCoin` implementation was dropped so it still incorrectly returns coins from the base view (but this doesn't matter because the method never gets called).

l0rinc commented:
I'm not sure I understand what is happening here.
There are many ways to do this cleanup, and based on the feedback, I have iterated through several versions.
Given that `PeekCoin` was just introduced (and is still unused in a few places, which is why we're doing the cleanup in the first place), and based on the feedback we're getting, it does seem like it's more complicated than it needs to be, which is fine. That's exactly what these cleanups are meant to address.
We have several follow-ups after this PR. It's fine if this one doesn't resolve every single problem.

> I find refactoring PRs the most exhausting to review

I split the change into tiny commits to help with that, and I applied the feedback I received on how to improve it. What is your exact suggestion here?

> better to minimise their costs and get them out of the way easily so that focus can go back to the PRs that are adding improvements directly

Agreed, so why are you NACKing this one exactly?

>  it's cleaning up a bunch of internal code ugliness, not improving bitcoin in ways that are directly detectable by users

> Noticing behavioural changes

Can you please tell me the exact steps that changed here? How does a user notice the "changes" you mention?

> should have resulted in an immediate "oops, yeah, that change should go in a followup, I'll open one now", not a debate

which is exactly what [happened](https://github.com/bitcoin/bitcoin/pull/34124#pullrequestreview-3922307527), but you "haven't looked at the updated PR"

sedited reviewed (approved):
ACK 1d1dc4ed2cfcc573e73ddef2d15b6e64df4f5ff3

w0xlt reviewed (commented):
reACK https://github.com/bitcoin/bitcoin/commit/1d1dc4ed2cfcc573e73ddef2d15b6e64df4f5ff3

l0rinc commented:
> I think you should fix it in this PR or open a separate PR that is merged before this one

Yes, that's what I meant. I haven't yet found a version that makes sense, since I don't think the backend should be mutable in the first place - it's hard to make this swapping fuzzer realistic.
I'll probably just go with a resetting the state on a swap for now.

l0rinc commented:
Trivial rebase, ready for review again.

l0rinc commented:
The new [FreeBSD Clang cross job](https://github.com/bitcoin/bitcoin/pull/34491) (maybe combined with some other compiler change) already paid off by catching a shared-library build warning here (cc: @fanquake), so I moved the `CoinsViewEmpty` singleton definition to `coins.cpp`.

ryanofsky reviewed (approved):
Code review ACK e161ca2c20d70dfb601d8290a026ded81e9cb769. Changes still look good to me.

Only changes since last review were rebasing, avoiding freebsd shared library warning, avoiding a bug in the fuzz test exposed by changes here. I did leave a readability suggestion below but it's not important.

w0xlt reviewed (commented):
reACK e161ca2c20d70dfb601d8290a026ded81e9cb769

w0xlt reviewed (commented):
reACK 8783cc8056db3adcec85be01683d716aea4d5130

ryanofsky reviewed (approved):
Code review ACK 8783cc8056db3adcec85be01683d716aea4d5130. Just updated comments and variable name since last review. The fuzz test code is much clearer now IMO

andrewtoth-exo reviewed (commented):
ACK 8783cc8056db3adcec85be01683d716aea4d5130

ajtowns commented on src/coins.h:
```diff
@@ -303,63 +303,89 @@ struct CoinsViewCacheCursor
 };
 
+/** Noop coins view. */
+class CoinsViewEmpty : public CCoinsView
```
I think this would probably be better as a `final` class -- the only things inheriting from it are the CCoinsViewTest and CoinsViewBottom test harnesses, which are just taking advantage of the default functions, which are probably better implemented directly in harnesses anyway (eg `EstimateSize()` probably shouldn't just `return 0` for non-empty coins views)

ajtowns reviewed (approved):
ACK 8783cc8056db3adcec85be01683d716aea4d5130

`CoinsViewEmpty` being final is a suggestion for future work, not here.

I think it would have been simpler to review this if this were a pure-refactor (code cleanup, whitespace, Empty singleton, CCoinsView pure virtual), with the functional changes (fuzz backend, `Assert(in_view)`) broken out as a separate PR. I think the advice in CONTRIBUTING.md, "In general, refactoring pull requests should not mix these
three kinds of activities in order to make refactoring pull requests easy to
review and uncontroversial." is perhaps misleading -- different kinds of refactoring should be split into different commits, but can still be mixed within a PR.

ryanofsky commented:
Note: ack from https://github.com/bitcoin/bitcoin/pull/34124#pullrequestreview-4086278658 does seem like a reack, not an impersonation https://github.com/bitcoin/bips/pull/2120#issuecomment-4055965484

l0rinc commented:
Yes, I can confirm it was @andrewtoth

andrewtoth commented:
ACK 8783cc8056db3adcec85be01683d716aea4d5130

Sorry for the confusion. @andrewtoth-exo is my alt account.

l0rinc commented:
Merged `PeekCoin` into `GetCoin` in https://github.com/bitcoin/bitcoin/pull/35078, as suggested by @ryanofsky and @ajtowns - thanks for the reviews and suggestions!
